### ISSUE TITLE
WebSocket connection to 'wss://api-such.andsuch.xyz/graphql/' failed: Error during WebSocket handshake: Unexpected response code: 400

### ISSUE THREAD

**Issue by Cimmanuel** [datadvance/DjangoChannelsGraphqlWs#46]
Hello @prokher, thanks for this great library. I recently deployed my application which uses [`django-graphql-playground`](https://github.com/jaydenwindle/django-graphql-playground) to test subscriptions. It works great locally. However in production, I get the following error in my browser console:

```
WebSocket connection to 'wss://api-such.andsuch.xyz/graphql/' failed: Error during WebSocket handshake: Unexpected response code: 400
```
...and in `graphql-playground`,  I get:

```
{
  "error": "Could not connect to websocket endpoint wss://api-such.andsuch.xyz/graphql/. Please check if the endpoint url is correct."
}
```
I checked Network tab in Chrome and discovered that websocket connections are closing immediately. What could be causing this?
Here's what my `docker-compose.yml` looks like:

```
version: '3.7'

services:
  nginx:
    container_name: nginx
    image: nginx
    restart: always
    depends_on:
      - web
    volumes:
      - ./web/dev.nginx.template:/etc/nginx/conf.d/dev.nginx.template
      - ./static:/static
      - ./media:/media
    ports:
      - "8080:80"
    networks:
      - SOME_NETWORK
    command: /bin/bash -c "envsubst \"`env | awk -F = '{printf \" $$%s\", $$1}'`\" < /etc/nginx/conf.d/dev.nginx.template > /etc/nginx/conf.d/default.conf && exec nginx -g 'daemon off;'"

  web:
    container_name: web
    restart: always
    build: ./web
    networks:
      - SOME_NETWORK
    depends_on:
      - postgres
      - redis
    volumes:
      - ./web:/usr/src/app/
    environment:
      - REDIS_HOST=redis
      - REDIS_PORT=6379
      - GRAPHQL_ENDPOINT=https://api-such.andsuch.xyz/graphql/
    command: bash -c /start.sh

  postgres:
    container_name: postgres
    restart: always
    image: postgres:latest
    networks:
      - SOME_NETWORK
    volumes:
      - pgdata:/var/lib/postgresql/data/

  redis:
   container_name: redis
   restart: always
   image: redis:latest
   networks:
    - SOME_NETWORK
   ports:
     - "6379:6379"
   volumes:
     - redisdata:/data

volumes:
  pgdata:
  redisdata:

networks:
  SOME_NETWORK:
    name: SOME_NETWORK
    driver: bridge
```

`settings.py`

```python
...
...
CHANNEL_LAYERS = {
    'default': {
        'BACKEND': 'channels_redis.core.RedisChannelLayer',
        'CONFIG': {
            'hosts': [(os.getenv('REDIS_HOST', 'redis'), os.getenv('REDIS_PORT', 6379))],
        }
    }
}
...
...
```

`urls.py`

```python
...
...
from graphql_playground.views import GraphQLPlaygroundView

urlpatterns = [
    path('admin/', admin.site.urls),
    path('playground/', GraphQLPlaygroundView.as_view(
        endpoint=os.getenv('GRAPHQL_ENDPOINT'))),
]
...
```

`nginx`

```
server {
    client_max_body_size 10M; 
    
    listen 443 ssl;
    listen [::]:443 ssl;
    
    server_name api-such.andsuch.xyz;
    
    ssl_certificate /etc/ssl/certs/andsuch.xyz.pem;
    ssl_certificate_key /etc/ssl/certs/andsuch.xyz.key;
    
    location = /favicon.ico { access_log off; log_not_found off; }
    
    location / {
        proxy_set_header Host $http_host;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection ‘upgrade’;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://0.0.0.0:8080;
    }
}
```

What could be wrong? I'm out of ideas

**Comment by psdon** [comment]
@Cimmanuel I guess you forgot to add `channels` to your installed_apps in settings config

**Comment by Cimmanuel** [comment]
@psdon I assure you that's not what's happening

**Comment by Cimmanuel** [comment]
After going through lots of articles, I discovered [this one](https://testdriven.io/blog/dockerizing-django-with-postgres-gunicorn-and-nginx/), and [this section](https://testdriven.io/blog/dockerizing-django-with-postgres-gunicorn-and-nginx/#nginx) made me realize the only thing that was wrong was the fact that I didn't expose port 8000 internally to other docker services. In my `docker-compose` file, the web service was supposed to have...

```
...
...
expose:
    - 8000
...
...
```
... in it. I added that and it was solved.
